lib/charms/layer/openstack.py

```diff
@@ -251,6 +252,30 @@ def _save_creds(creds_data):
     kv().set('charm.openstack.full-creds', creds_data)
 
 
+def get_creds_and_reformat():
+    creds = _load_creds()
+    formatted_creds = {}
+    auth_url_parsed = urlparse(creds['auth_url'])
+
+    formatted_creds['auth_protocol'] = auth_url_parsed.scheme
+    formatted_creds['credentials_protocol'] = auth_url_parsed.scheme
+    formatted_creds['auth_host'] = auth_url_parsed.hostname
+    formatted_creds['credentials_host'] = auth_url_parsed.hostname
+    formatted_creds['auth_port'] = auth_url_parsed.port
+    formatted_creds['credentials_port'] = auth_url_parsed.port
+    formatted_creds['api_version'] = creds['version']
+    formatted_creds['credentials_username'] = creds['username']
+    formatted_creds['credentials_password'] = creds['password']
+    formatted_creds['credentials_project'] = creds['project_name']
+    formatted_creds['credentials_project_id'] = '1'
```

Comment:
auth_data_complete() assumes project_id is necessary, however we don't have it from juju trust
https://github.com/openstack/charm-interface-keystone-credentials/blob/master/requires.py#L92

Reply:
Hrm.  While we don't pull it out in `_normalize_creds`, my test environment at least does include the project ID (called `tenant-id` there, though) in the credentials returned from Juju, so for completeness it might be worth also pulling that out, if it's available.

I think the reason that the relation considers the project ID required is that it doesn't return the ID of the project domain and the caller quite possibly doesn't have it and may not even have the domain name if they're using the default (unless it's always actually called "Default"?).  Also note that just the domain name may not be sufficient:

```shell
ubuntu@k8steam-bastion-2:~$ openstack project show --domain $OS_PROJECT_DOMAIN_NAME $OS_PROJECT_NAME -c id -f value
You are not authorized to find domain with the name 'user'.
ubuntu@k8steam-bastion-2:~$ openstack project show --domain $OS_PROJECT_DOMAIN_ID $OS_PROJECT_NAME -c id -f value
b0f70593c2e14ea1a412ab3636ca2136
```

However, it's also worth noting that, for this particular use-case, k8s-master doesn't actually use any of the auth data (or the SSL data, either, actually, since it oddly gets the CA from a config option rather than the relation, for some reason), so it really should be using the `keystone-credentials.available` flag rather than the `keystone-credentials.available.auth` flag, anyway.  It's probably worth fixing that there so that we don't run into an issue where one or more unused values aren't available for whatever reason and block things unnecessarily, but that's a separate issue, which I can file a bug for.

Reply:
As mentioned in chat, I'm also fine just saying that this charm only partially supports this interface, dropping all the auth data entirely, since it can't actually create a new Keystone user as requested, and the actual use-case for this means that partial support is sufficient along with https://github.com/charmed-kubernetes/charm-kubernetes-master/pull/127

Reply:
@johnsca removed auth data please have a look
https://github.com/juju-solutions/charm-openstack-integrator/pull/38/commits/c11cb7e36b01e386035043861b3d1bef6dceb37b